scripts/buf.pl

```diff
@@ -40,64 +43,59 @@
-    umask $old_umask;
-
-    print $fh join("\0", map $_->{server}->{address} . $_->{name}, channels), "\n";
+    my $out = { suppress => join("\0", map $_->{server}->{address} . $_->{name}, channels) };
     for my $window (windows) {
-	next unless defined $window;
-	next if $window->{name} eq 'status';
-	my $view = $window->view;
-	my $line = $view->get_lines;
-	my $lines  = 0;
-	my $buf = '';
-	if (defined $line){
-	    {
-		$buf .= $line->get_text(1) . "\n";
-		$line = $line->next;
-		$lines++;
-		redo if defined $line;
-	    }
-	}
-	printf $fh "%s:%s\n%s", $window->{refnum}, $lines, $buf;
+        next unless defined $window;
+        next if $window->{name} eq 'status';
+        my $view = $window->view;
+        my $line = $view->get_lines;
+        my $lines  = 0;
+        my $buf = '';
+        my $output;
+        if (defined $line) {
+            {
+              push @$output, { level => $line->{info}{level}, data => $line->get_text(1) };
+              $line = $line->next;
+              redo if defined $line;
+            }
+        }
+        push @{$out->{windows}}, { refnum => $window->{refnum}, lines => $output };
     }
-    close $fh;
+    my $old_umask = umask 0077;
+    store($out, _filename) or die $!;
+    umask $old_umask;
     unlink sprintf("%s/sessionconfig", get_irssi_dir);
     command 'layout save';
     command 'save';
 }
 
 sub restore {
     my $fn = _filename;
-    open my $fh, q{<}, $fn or die "open $fn: $!";
-    unlink $fn or warn "unlink $fn: $!";
-
-    my @suppress = split /\0/, readline $fh;
-    if (settings_get_bool 'upgrade_suppress_join') {
-	chomp $suppress[-1];
-	@suppress{@suppress} = (2) x @suppress;
-    }
+    my $in = retrieve($fn) or die $!;
+    unlink $fn;
+  
+    my @suppress = split /\0/, $in->{suppress};
+    @suppress{@suppress} = (2) x @suppress if (settings_get_bool 'upgrade_suppress_join');
+  
     active_win->command('^window scroll off');
-    while (my $bla = readline $fh){
-	chomp $bla;
-	my ($refnum, $lines) = split /:/, $bla;
-	next unless $lines;
-	my $window = window_find_refnum $refnum;
-	unless (defined $window){
-	    readline $fh for 1..$lines;
-	    next;
-	}
-	my $view = $window->view;
-	$view->remove_all_lines();
-	$view->redraw();
-	my $buf = '';
-	$buf .= readline $fh for 1..$lines;
-	my $sep = settings_get_str 'upgrade_separator';
-	$sep .= "\n" if $sep ne '';
-	$window->gui_printtext_after(undef, MSGLEVEL_CLIENTNOTICE, "$buf\cO$sep");
-	$view->redraw();
+    for my $win (@{$in->{windows}}) {
+        my $window = window_find_refnum $win->{refnum};
+        next unless $window;
+        my @lines  = @{ $win->{lines} || [] };
+        next unless @lines;
+    
+        my $view = $window->view;
+        $view->remove_all_lines();
+        $view->redraw();
+        my $sep = settings_get_str 'upgrade_separator';
+        $sep .= "\n" if $sep ne '';
+        $window->gui_printtext_after(undef, MSGLEVEL_CLIENTNOTICE, "\cO$sep\n");
```

Comment:
this separator notice should go LAST after all the other lines have been printed in each window, at least that's how it originally was

Reply:
iirc the screen got scrolled up if I did it that way so the =Upgrade= wasn't visible

Reply:
move it below the loop and use $window->last_line_insert instead of undef, should work I think